src/ripple/ledger/tests/View_test.cpp

```diff
@@ -524,6 +525,209 @@ class View_test
+        auto const EUR = gw["EUR"];
+
+        env.fund(XRP(10000), alice, bob, carol, gw);
+        env.trust(USD(100), alice, bob, carol);
+        {
+            // Global freezing.
+            env(pay(gw, alice, USD(50)));
+            env(offer(alice, XRP(5), USD(5)));
+
+            // Now freeze gw.
+            env(fset (gw, asfGlobalFreeze));
+            env.close();
+            env(offer(alice, XRP(4), USD(5)), ter(tecFROZEN));
+            env.close();
+
+            // Alice's USD balance should be zero if frozen.
+            expect(USD(0) == accountHolds (*env.closed(),
+                alice, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
+
+            // Thaw gw and try again.
+            env(fclear (gw, asfGlobalFreeze));
+            env.close();
+            env(offer("alice", XRP(4), USD(5)));
+        }
+        {
+            // Local freezing.
+            env(pay(gw, bob, USD(50)));
+            env.close();
+
+            // Now gw freezes bob's USD trust line.
+            env(trust(gw, USD(100), bob, tfSetFreeze));
+            env.close();
+
+            // Bob's balance should be zero if frozen.
+            expect(USD(0) == accountHolds (*env.closed(),
+                bob, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
+
+            // gw thaws bob's trust line.  bob gets his money back.
+            env(trust(gw, USD(100), bob, tfClearFreeze));
+            env.close();
+            expect(USD(50) == accountHolds (*env.closed(),
+                bob, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
+        }
+        {
+            // accountHolds().
+            env(pay(gw, carol, USD(50)));
+            env.close();
+
+            // carol has no EUR.
+            expect(EUR(0) == accountHolds (*env.closed(),
+                carol, EUR.currency, gw, fhZERO_IF_FROZEN, env.journal));
+
+            // But carol does have USD.
+            expect(USD(50) == accountHolds (*env.closed(),
+                carol, USD.currency, gw, fhZERO_IF_FROZEN, env.journal));
+
+            // carol's XRP balance should be her holdings minus her reserve.
+            auto const carolsXRP = accountHolds (*env.closed(),
+                carol, xrpCurrency(), gw, fhZERO_IF_FROZEN, env.journal);
+            expect(carolsXRP == XRP(9750));
+
+            // carol should be able to spend *more* than her XRP balance on
+            // a fee by eating into her reserve.
+            env(noop(carol), fee(carolsXRP + XRP(10)));
+            env.close();
+
+            // carol's XRP balance should now show as zero.
+            expect (XRP(0) == accountHolds (*env.closed(),
+                carol, xrpCurrency(), gw, fhZERO_IF_FROZEN, env.journal));
+        }
+        {
+            // accountFunds().
+            // Gateways have whatever funds they claim to have.
+            auto const gwUSD = accountFunds(
+                *env.closed(), gw, USD(0), fhZERO_IF_FROZEN, env.journal);
+            expect (gwUSD == USD(0));
```

Comment:
Since the gateway can have whatever they claim to have, we should test a non-zero default as well.